### PR TITLE
gh-118465: Add __firstlineno__ attribute to class

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
It is set by compiler with the line number of the first line of the class definition.

* Issue: gh-118465

----
📚 Documentation preview 📚: https://cpython-previews--118475.org.readthedocs.build/

### CODE REVIEW

File: Doc/reference/datamodel.rst [hunk 1]
```diff
@@ -1005,6 +1006,9 @@ Special attributes:
       A tuple containing names of attributes of this class which are accessed
       through ``self.X`` from any function in its body.
 
+   :attr:`__firstlineno__`
+      The line number of the first line of the xlass definition.
```
Comment (markshannon) [inline Doc/reference/datamodel.rst]:
```suggestion
      The line number of the first line of the class definition.
```

File: Doc/whatsnew/3.13.rst [hunk 2]
```diff
@@ -280,6 +280,11 @@ Other Language Changes
   class scopes are not inlined into their parent scope. (Contributed by
   Jelle Zijlstra in :gh:`109118` and :gh:`118160`.)
 
+* Classes have a new :attr:`!__firstlineno__` attribute,
+  populated by the compiler, with the line number of the first line
+  of the class definition.
```
Comment (iritkatriel) [inline Doc/whatsnew/3.13.rst:300]: If the class has decorators then I think this is the first line of the first decorator. Need a test for this case, and probably to mention in the doc.

Comment (serhiy-storchaka) [reply Doc/whatsnew/3.13.rst:300]: There are existing tests, and they are passed.

I'll add explicit mentioning of this fact, but AFAIK it was not specified for `co_firstlineno` etc.

File: Doc/reference/datamodel.rst [hunk 3]
```diff
@@ -1005,6 +1006,9 @@ Special attributes:
       A tuple containing names of attributes of this class which are accessed
       through ``self.X`` from any function in its body.
 
+   :attr:`__firstlineno__`
+      The line number of the first line of the class definition.
```
Comment (serhiy-storchaka) [inline Doc/reference/datamodel.rst]:
```suggestion
      The line number of the first line of the class definition, including decorators.
```

**Comment by JelleZijlstra** [comment]
It is possible to construct classes without `__firstlineno__`:

```python
def f():
    __firstlineno__ = 1
    class X:
        nonlocal __firstlineno__
    return X
print(f().__firstlineno__)  # AttributeError
```

Or:

```python
class Y:
    global __firstlineno__
print(Y.__firstlineno__)  # AttributeError
```

This is obviously an extreme edge case though. You could also do this by manipulating `locals()` in the class body or the metaclass.

Is it worth it to make `inspect` smarter in the presence of classes without `__firstlineno__`?

```
  File "/Users/jelle/py/cpython/insp.py", line 18, in <module>
    print(inspect.findsource(X))
          ~~~~~~~~~~~~~~~~~~^^^
  File "/Users/jelle/py/cpython/Lib/inspect.py", line 1070, in findsource
    return lines, object.__firstlineno__ - 1
                  ^^^^^^^^^^^^^^^^^^^^^^
AttributeError: type object 'X' has no attribute '__firstlineno__'
```

**Comment by serhiy-storchaka** [comment]
Good point. I thought that it is the same as for `__module__` and `__qualname__`, but it turned out that they are set implicitly in the `type` constructor even when not specified in the class dict. They are set even for `type('A', (), {})`.

We can try to set `__firstlineno__` to the current line number in the top Python frame, but it is safer to not set it at all. Setting `__firstlineno__` for classes created by the `type()` call rather than in the `class` statement does not help in the only use case for this attribute.

I think that for backward compatibility `getsource()` should convert an AttributeError into OSError which was raised before for classes created by the `type()` call.

**Review by JelleZijlstra — APPROVED** [review]

**Pull request merged**